internal/providers/msentraid/msentraid_test.go

```diff
@@ -177,3 +167,108 @@ func TestGetUserInfo(t *testing.T) {
 		})
 	}
 }
+
+func TestIsTokenForDeviceRegistration(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		appID        string
+		invalidToken bool
+
+		want    bool
+		wantErr bool
+	}{
+		"Success_when_token_has_microsoft_broker_app_ID": {appID: consts.MicrosoftBrokerAppID, want: true},
+		"Success_when_token_has_other_app_ID":            {appID: "some-other-app-id", want: false},
+		"Success_when_token_has_empty_app_ID":            {appID: "", want: false},
+
+		"Error_when_token_has_no_app_ID": {appID: "-", wantErr: true},
+		"Error_when_token_is_invalid":    {invalidToken: true, wantErr: true},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			claims := jwt.MapClaims{"appid": tc.appID}
+			if tc.appID == "-" {
+				claims = jwt.MapClaims{}
+			}
+
+			accessToken := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
+			accessTokenString, err := accessToken.SignedString(testutils.MockKey)
+			require.NoError(t, err, "Failed to sign access token")
+
+			if tc.invalidToken {
+				accessTokenString = "invalid-token"
+			}
+
+			token := &oauth2.Token{AccessToken: accessTokenString}
+
+			p := msentraid.New()
+			got, err := p.IsTokenForDeviceRegistration(token)
+
+			if tc.wantErr {
+				require.Error(t, err, "IsTokenForDeviceRegistration should return an error")
+				return
+			}
+			require.NoError(t, err, "IsTokenForDeviceRegistration should not return an error")
+			require.Equal(t, tc.want, got, "IsTokenForDeviceRegistration should return the expected value")
+		})
+	}
+}
+
+func TestMaybeRegisterDevice(t *testing.T) {
+	// Registering a device is not thread-safe, because libhimmelblau sets BrokerClientApplication.cert_key
+	// after successful device registration, and uses the same field in subsequent calls,
+	// e.g. in the acquire_token_by_refresh_token method. If the cert_key does not match the expected value,
+	// libhimmelblau returns "TPM error: Failed to load IdentityKey: Aes256GcmDecrypt".
+	mockMSServerRegisterDeviceMu.Lock()
+	defer mockMSServerRegisterDeviceMu.Unlock()
+
+	t.Parallel()
+
+	maybeRegisterDevice(t)
```

Comment:
It would be nice if you could add some concurrent tests here, since I feel this is not playing well in such case

Reply:
Or well, once the mock is setup, try to use the machinery to register multiple devices in parallel and see that we're not failing